File: sdk-python/src/agentvend_sdk/validation_client.py
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Union

from .agentvend_headers import AgentVendHeaders
from .hmac_utils import validate_hmac_signature

DEFAULT_CORE_PATH_PREFIX = "/core/api/v1"


def _core_agent_keys_url(base_url: str, core_path_prefix: Optional[str], suffix: str) -> str:
    base = base_url.rstrip("/")
    p = (core_path_prefix or DEFAULT_CORE_PATH_PREFIX).strip()
    if not p.startswith("/"):
        p = "/" + p
    p = p.rstrip("/")
    return f"{base}{p}/agent-keys/{suffix}"


def _validate_url(base_url: str, core_path_prefix: Optional[str]) -> str:
    return _core_agent_keys_url(base_url, core_path_prefix, "validate")


@dataclass
class AgentKeyValidationResult:
    user_id: Optional[str]
    agent_id: Optional[str]
    plan: Optional[str]
    roles: List[str]
    quota_remaining: Optional[float]
    subscription_active: bool
    billing_model_type: Optional[str]
    measurement_type: Optional[str]
    unit_label: Optional[str]


@dataclass
class UsageEstimateResult:
    sufficient_credits: bool
    would_exceed_cap: bool
    would_allow: bool
    estimated_cost: Optional[float]
    remaining_credits: Optional[float]
    remaining_spending_cap: Optional[float]
    billing_model_type: Optional[str]
    measurement_type: Optional[str]
    unit_label: Optional[str]
    breakdown: Optional[Dict[str, Any]]
    estimate_schema_version: int
    timestamp: int
    http_status: int


def validate_agent_key(
    base_url: str,
    agent_key: str,
    agent_secret: str,
    agent_id: Optional[str] = None,
    *,
    core_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> Optional[AgentKeyValidationResult]:
    """Validate agent key via Core service. Requires 'requests' (pip install requests)."""
    try:
        import requests
    except ImportError:
        raise ImportError("validate_agent_key requires 'requests'. pip install requests")
    url = _validate_url(base_url, core_path_prefix)
    body = {"agentKey": agent_key, "agentId": agent_id, "agentSecret": agent_secret}
    sess = session or requests.Session()
    resp = sess.post(url, json=body)
    if not resp.ok:
        return None
    response_text = resp.text
    signature = resp.headers.get(AgentVendHeaders.SIGNATURE)
    timestamp = resp.headers.get(AgentVendHeaders.TIMESTAMP)
    if not signature or not timestamp:
        return None
    if not validate_hmac_signature(signature, response_text + timestamp, agent_secret):
        return None
    data = resp.json()
    if not data.get("valid"):
        return None
    roles = data.get("roles") or []
    return AgentKeyValidationResult(
        user_id=data.get("userId"),
        agent_id=data.get("agentId") or agent_id,
        plan=data.get("plan"),
        roles=roles if isinstance(roles, list) else [],
        quota_remaining=data.get("quotaRemaining"),
        subscription_active=bool(data.get("subscriptionActive")),
        billing_model_type=data.get("billingModelType"),
        measurement_type=data.get("measurementType"),
        unit_label=data.get("unitLabel"),
    )


def estimate_usage(
    base_url: str,
    agent_key: str,
    agent_secret: str,
    estimated_units: Union[int, float],
    agent_id: Optional[str] = None,
    *,
    core_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> Optional[UsageEstimateResult]:
    """Usage pre-flight via Core (same trust model as :func:`validate_agent_key`). Requires ``requests``."""
    try:
        import requests
    except ImportError:
        raise ImportError("estimate_usage requires 'requests'. pip install requests")
    if estimated_units is None or float(estimated_units) <= 0:
        return None
    if not agent_key or not str(agent_key).strip():
        return None
    url = _core_agent_keys_url(base_url, core_path_prefix, "estimate-usage")
    body = {
        "agentKey": agent_key,
        "agentId": agent_id,
        "agentSecret": agent_secret,
        "estimatedUnits": estimated_units,
    }
    sess = session or requests.Session()
    resp = sess.post(url, json=body)
    code = resp.status_code
    if code not in (200, 403, 429):
        return None
    response_text = resp.text
    if not response_text.strip():
        return None
    signature = resp.headers.get(AgentVendHeaders.SIGNATURE)
    timestamp = resp.headers.get(AgentVendHeaders.TIMESTAMP)
    if not signature or not timestamp:
        return None
    if not validate_hmac_signature(signature, response_text + timestamp, agent_secret):
        return None
    data = resp.json()
    breakdown = data.get("breakdown")
    if breakdown is not None and not isinstance(breakdown, dict):
        breakdown = None
    return UsageEstimateResult(
        sufficient_credits=bool(data.get("sufficientCredits")),
        would_exceed_cap=bool(data.get("wouldExceedCap")),
        would_allow=bool(data.get("wouldAllow")),
        estimated_cost=data.get("estimatedCost"),
        remaining_credits=data.get("remainingCredits"),
        remaining_spending_cap=data.get("remainingSpendingCap"),
        billing_model_type=data.get("billingModelType"),
        measurement_type=data.get("measurementType"),
        unit_label=data.get("unitLabel"),
        breakdown=breakdown,
        estimate_schema_version=int(data.get("estimateSchemaVersion", 0)),
        timestamp=int(data.get("timestamp", 0)),
        http_status=code,
    )

File: sdk-dotnet/ValidationClient.cs
using System.Net;
using System.Net.Http;
using System.Text;
using System.Text.Json;

namespace Tollara;

public record ServiceKeyValidationResult(
    string? UserId,
    string? ServiceId,
    string? ServiceProductId,
    IReadOnlyList<string> Roles,
    string? SubscriptionStatus,
    int? ValidationSchemaVersion,
    string? BillingModelType,
    string? MeasurementType,
    string? UnitLabel,
    Guid? ServiceKeyId)
{
    public bool GrantAccess() => Verifier.GrantAccess(SubscriptionStatus);

    public static bool GrantAccess(string? subscriptionStatus) => Verifier.GrantAccess(subscriptionStatus);
}

/// <summary>Wire result for Core usage estimate endpoints (see docs-sdk/MAIN-SDK-API-SPEC.md §2.3).</summary>
public record UsageEstimateResult(
    bool SufficientCredits,
    bool WouldExceedCap,
    bool WouldAllow,
    decimal? EstimatedCost,
    string? BillingModelType,
    string? MeasurementType,
    string? UnitLabel,
    UsageBreakdown? Breakdown,
    int EstimateSchemaVersion,
    long Timestamp,
    int HttpStatus);

/// <summary>Canonical failure codes for validate outcome (§2.1.1).</summary>
public enum ValidationFailureCode
{
    MISSING_KEY,
    NETWORK,
    HTTP_ERROR,
    MISSING_SIGNATURE_HEADERS,
    HMAC_MISMATCH,
    INVALID_KEY,
    PARSE_ERROR,
}

public record ServiceKeyValidationFailure(
    ValidationFailureCode Code,
    string? Message = null,
    int? HttpStatus = null);

public abstract record ServiceKeyValidationOutcome
{
    public sealed record Success(ServiceKeyValidationResult Result) : ServiceKeyValidationOutcome;
    public sealed record Failure(ServiceKeyValidationFailure Error) : ServiceKeyValidationOutcome;

    public bool Ok => this is Success;
}

public static class ValidationClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    /// <summary>Validate using the SDK default API origin and Core path prefix (same as <see cref="TollaraClient"/>).</summary>
    public static Task<ServiceKeyValidationResult?> ValidateServiceKeyAsync(HttpClient http, string serviceKey, string? serviceId,
        string serviceSecret, CancellationToken ct = default) =>
        ValidateServiceKeyAsync(http, DefaultCoreServiceRoot(), serviceKey, serviceId, serviceSecret, ct);

    /// <summary>Validate using the SDK default API origin and Core path prefix (same as <see cref="TollaraClient"/>).</summary>
    public static Task<ServiceKeyValidationResult?> ValidateServiceKeyAsync(HttpClient http, string serviceKey, string? serviceId,
        string serviceSecret, CancellationToken ct = default) =>
        ValidateServiceKeyAsync(http, DefaultCoreServiceRoot(), serviceKey, serviceId, serviceSecret, ct);

    /// <summary>Validate with structured outcome (§2.1.1).</summary>
    public static Task<ServiceKeyValidationOutcome> ValidateServiceKeyWithOutcomeAsync(HttpClient http, string serviceKey,
        string? serviceId, string serviceSecret, CancellationToken ct = default) =>
        ValidateServiceKeyWithOutcomeAsync(http, DefaultCoreServiceRoot(), serviceKey, serviceId, serviceSecret, ct);

    /// <summary>Validate against an explicit Core service base (include path prefix, e.g. custom or local stack).</summary>
    public static async Task<ServiceKeyValidationOutcome> ValidateServiceKeyWithOutcomeAsync(HttpClient http, string coreServiceUrl,
        string serviceKey, string? serviceId, string serviceSecret, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(serviceKey))
            return new ServiceKeyValidationOutcome.Failure(new ServiceKeyValidationFailure(ValidationFailureCode.MISSING_KEY));

        var url = coreServiceUrl.TrimEnd('/') + "/service-keys/validate";
        var body = new { serviceKey, serviceId, serviceSecret };
        var bodyStr = JsonSerializer.Serialize(body);
        var req = new HttpRequestMessage(HttpMethod.Post, url) { Content = new StringContent(bodyStr, Encoding.UTF8, "application/json") };
        HttpResponseMessage res;
        try
        {
            res = await http.SendAsync(req, ct);
        }
        catch (HttpRequestException)
        {
            return new ServiceKeyValidationOutcome.Failure(new ServiceKeyValidationFailure(ValidationFailureCode.NETWORK));
        }

        var httpStatus = (int)res.StatusCode;
        if (!res.IsSuccessStatusCode)
            return new ServiceKeyValidationOutcome.Failure(
                new ServiceKeyValidationFailure(ValidationFailureCode.HTTP_ERROR, HttpStatus: httpStatus));

        var responseText = await res.Content.ReadAsStringAsync(ct);
        var signature = res.Headers.TryGetValues(TollaraHeaders.Signature, out var sig) ? string.Join("", sig) : null;
        var timestamp = res.Headers.TryGetValues(TollaraHeaders.Timestamp, out var ts) ? string.Join("", ts) : null;
        if (string.IsNullOrEmpty(signature) || string.IsNullOrEmpty(timestamp))
            return new ServiceKeyValidationOutcome.Failure(
                new ServiceKeyValidationFailure(ValidationFailureCode.MISSING_SIGNATURE_HEADERS, HttpStatus: httpStatus));

        if (!Hmac.ValidateHmacWithTimestamp(signature, responseText, timestamp, serviceSecret))
            return new ServiceKeyValidationOutcome.Failure(
                new ServiceKeyValidationFailure(ValidationFailureCode.HMAC_MISMATCH, HttpStatus: httpStatus));

        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(responseText);
        }
        catch (JsonException)
        {
            return new ServiceKeyValidationOutcome.Failure(
                new ServiceKeyValidationFailure(ValidationFailureCode.PARSE_ERROR, HttpStatus: httpStatus));
        }

        var root = doc.RootElement;
        if (root.TryGetProperty("valid", out var v) && !v.GetBoolean())
        {
            var message = ReadString(root, "error");
            return new ServiceKeyValidationOutcome.Failure(
                new ServiceKeyValidationFailure(ValidationFailureCode.INVALID_KEY, message, httpStatus));
        }

        var roles = root.TryGetProperty("roles", out var r) ? r.EnumerateArray().Select(x => x.GetString() ?? "").ToList() : new List<string>();
        Guid? serviceKeyId = null;
        if (root.TryGetProperty("serviceKeyId", out var sk) && sk.ValueKind == JsonValueKind.String &&
            Guid.TryParse(sk.GetString(), out var skGuid))
            serviceKeyId = skGuid;

        var result = new ServiceKeyValidationResult(
            root.TryGetProperty("userId", out var uid) ? uid.GetString() : null,
            root.TryGetProperty("serviceId", out var sid) ? sid.GetString() : serviceId,
            ReadString(root, "serviceProductId"),
            roles,
            ReadString(root, "subscriptionStatus"),
            root.TryGetProperty("validationSchemaVersion", out var vsv) && vsv.TryGetInt32(out var vv) ? vv : null,
            ReadString(root, "billingModelType"),
            ReadString(root, "measurementType"),
            ReadString(root, "unitLabel"),
            serviceKeyId
        );
        return new ServiceKeyValidationOutcome.Success(result);
    }

    /// <summary>Validate against an explicit Core service base (include path prefix, e.g. custom or local stack).</summary>
    public static async Task<ServiceKeyValidationResult?> ValidateServiceKeyAsync(HttpClient http, string coreServiceUrl,
        string serviceKey, string? serviceId, string serviceSecret, CancellationToken ct = default)
    {
        var outcome = await ValidateServiceKeyWithOutcomeAsync(http, coreServiceUrl, serviceKey, serviceId, serviceSecret, ct);
        return outcome is ServiceKeyValidationOutcome.Success s ? s.Result : null;
    }

    /// <summary>Estimate usage using the SDK default Core root (same as <see cref="TollaraClient"/>).</summary>
    public static Task<UsageEstimateResult?> EstimateUsageAsync(HttpClient http, string serviceKey, decimal estimatedUnits,
        string? serviceId, string serviceSecret, CancellationToken ct = default) =>
        EstimateUsageAsync(http, DefaultCoreServiceRoot(), serviceKey, estimatedUnits, serviceId, serviceSecret, ct);

    /// <summary>POST estimate-usage to an explicit Core base (including path prefix).</summary>
    public static async Task<UsageEstimateResult?> EstimateUsageAsync(HttpClient http, string coreServiceUrl,
        string serviceKey, decimal estimatedUnits, string? serviceId, string serviceSecret, CancellationToken ct = default)
    {
        if (estimatedUnits <= 0) return null;
        if (string.IsNullOrWhiteSpace(serviceKey)) return null;
        var url = coreServiceUrl.TrimEnd('/') + "/service-keys/estimate-usage";
        var body = new { serviceKey, serviceId, serviceSecret, estimatedUnits };
        var bodyStr = JsonSerializer.Serialize(body);
        var req = new HttpRequestMessage(HttpMethod.Post, url) { Content = new StringContent(bodyStr, Encoding.UTF8, "application/json") };
        var res = await http.SendAsync(req, ct);
        var code = (int)res.StatusCode;
        if (code != (int)HttpStatusCode.OK && code != (int)HttpStatusCode.Forbidden &&
            code != (int)HttpStatusCode.TooManyRequests) return null;
        var responseText = await res.Content.ReadAsStringAsync(ct);
        if (string.IsNullOrWhiteSpace(responseText)) return null;
        var signature = res.Headers.TryGetValues(TollaraHeaders.Signature, out var sig) ? string.Join("", sig) : null;
        var timestamp = res.Headers.TryGetValues(TollaraHeaders.Timestamp, out var ts) ? string.Join("", ts) : null;
        if (string.IsNullOrEmpty(signature) || string.IsNullOrEmpty(timestamp)) return null;
        if (!Hmac.ValidateHmacWithTimestamp(signature, responseText, timestamp, serviceSecret)) return null;
        return ParseEstimateResult(responseText, code);
    }

    /// <summary>Core JWT usage estimate (§2.2). Response is not HMAC-signed.</summary>
    public static Task<UsageEstimateResult?> EstimateUsageWithJwtAsync(HttpClient http, string bearerToken, string userId,
        string serviceId, decimal estimatedUnits, CancellationToken ct = default) =>
        EstimateUsageWithJwtAsync(http, DefaultCoreServiceRoot(), bearerToken, userId, serviceId, estimatedUnits, ct);

    /// <summary>POST <c>…/billing/usage/estimate</c> with explicit Core root (including path prefix).</summary>
    public static async Task<UsageEstimateResult?> EstimateUsageWithJwtAsync(HttpClient http, string coreServiceUrl,
        string bearerToken, string userId, string serviceId, decimal estimatedUnits, CancellationToken ct = default)
    {
        if (estimatedUnits <= 0) return null;
        if (string.IsNullOrWhiteSpace(bearerToken) || string.IsNullOrWhiteSpace(userId) || string.IsNullOrWhiteSpace(serviceId))
            return null;
        var url = coreServiceUrl.TrimEnd('/') + "/billing/usage/estimate";
        var body = new { userId, serviceId, estimatedUnits };
        var bodyStr = JsonSerializer.Serialize(body);
        var req = new HttpRequestMessage(HttpMethod.Post, url) { Content = new StringContent(bodyStr, Encoding.UTF8, "application/json") };
        req.Headers.TryAddWithoutValidation("Authorization", "Bearer " + bearerToken.Trim());
        var res = await http.SendAsync(req, ct);
        var code = (int)res.StatusCode;
        if (code != (int)HttpStatusCode.OK && code != (int)HttpStatusCode.Forbidden &&
            code != (int)HttpStatusCode.TooManyRequests) return null;
        var responseText = await res.Content.ReadAsStringAsync(ct);
        if (string.IsNullOrWhiteSpace(responseText)) return null;
        return ParseEstimateResult(responseText, code);
    }

    private static UsageEstimateResult? ParseEstimateResult(string responseText, int httpStatus)
    {
        using var doc = JsonDocument.Parse(responseText);
        var root = doc.RootElement;
        UsageBreakdown? breakdown = null;
        if (root.TryGetProperty("breakdown", out var br) && br.ValueKind == JsonValueKind.Object)
            breakdown = JsonSerializer.Deserialize<UsageBreakdown>(br.GetRawText(), JsonOptions);
        return new UsageEstimateResult(
            root.TryGetProperty("sufficientCredits", out var sc) && sc.GetBoolean(),
            root.TryGetProperty("wouldExceedCap", out var wec) && wec.GetBoolean(),
            root.TryGetProperty("wouldAllow", out var wa) && wa.GetBoolean(),
            ReadDecimal(root, "estimatedCost"),
            ReadString(root, "billingModelType"),
            ReadString(root, "measurementType"),
            ReadString(root, "unitLabel"),
            breakdown,
            root.TryGetProperty("estimateSchemaVersion", out var esv) && esv.TryGetInt32(out var ev) ? ev : 0,
            root.TryGetProperty("timestamp", out var tsv) && tsv.TryGetInt64(out var tl) ? tl : 0L,
            httpStatus
        );
    }

    private static decimal? ReadDecimal(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var p)) return null;
        if (p.ValueKind == JsonValueKind.Null) return null;
        return p.ValueKind == JsonValueKind.Number && p.TryGetDecimal(out var d) ? d : null;
    }

    private static string? ReadString(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var p)) return null;
        if (p.ValueKind == JsonValueKind.Null) return null;
        return p.ValueKind == JsonValueKind.String ? p.GetString() : null;
    }

    private static string DefaultCoreServiceRoot() =>
        JoinBaseAndPrefix(TollaraClient.DefaultApiUrl, TollaraClient.DefaultCorePathPrefix);

    private static string JoinBaseAndPrefix(string baseUrl, string pathPrefix)
    {
        var b = baseUrl.Trim().TrimEnd('/');
        if (string.IsNullOrEmpty(pathPrefix)) return b;
        var p = pathPrefix.StartsWith('/') ? pathPrefix : "/" + pathPrefix;
        return b + p;
    }
}
